psi use case sample

snedamle opened the pull request.

snedamle committed:
psi use case sample

snedamle committed:
removing unnecessary files

snedamle committed:
removing unnecessary files

snedamle committed:
removing unnecessary files

rusheb commented on psi-sample/build.gradle:
```diff
@@ -0,0 +1,33 @@
+    id 'idea'
+}
+
+idea {
+    module {
+        downloadJavadoc = true
+    }
+}
```
Would it work to move this to `allprojects` so we don't have to repeat it inside and outside of `subprojects`?

snedamle replied:
as of now main conclave-samples is not a gradle project. We will change this for 1.1 when we update all the samples.

rusheb commented on psi-sample/client/build.gradle:
```diff
@@ -0,0 +1,14 @@
+dependencies {
+    implementation project(":common");
+    implementation "com.r3.conclave:conclave-client:$conclaveVersion"
+
```
Don't need this whitespace

snedamle replied:
Done

rusheb commented on psi-sample/common/build.gradle:
```diff
@@ -0,0 +1,16 @@
+plugins {
```
I can't comment on it directly, but I see we have a number of `.DS_Store` files. I think these are generated to finder and should be deleted and added to the global `.gitignore`

snedamle replied:
done

rusheb commented on psi-sample/common/build.gradle:
```diff
@@ -0,0 +1,16 @@
+    id 'java'
+}
+
+version 'unspecified'
```
Is it ok to use `unspecified` version?

snedamle replied:
I have removed this line. It is not required.

rusheb commented on psi-sample/common/build.gradle:
```diff
@@ -0,0 +1,16 @@
+
+version 'unspecified'
+
+sourceCompatibility = 1.8
```
Ah interesting, so you only need `sourceCompatibility` here and not `targetCompatibility`? I don't understand these parameters very well. Can you explain what it does?

snedamle replied:
sourceCompatibility Java version compatibility to use when compiling Java source. 

targetCompatibility is the Java version to generate classes for. The default value for targetCompatibility  is sourceCompatibility and hence that's why we dont need to specify targetCompatibility.
If you do not specify sourceCompatibility I get below error

No matching variant of project :common was found. The consumer was configured to find a runtime of a library compatible with Java 8, packaged as a jar, and its dependencies declared externally but:
          - Variant 'apiElements' capability overlay:common:unspecified declares a library, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares an API of a component compatible with Java 11 and the consumer needed a runtime of a component compatible with Java 8
          - Variant 'runtimeElements' capability overlay:common:unspecified declares a runtime of a library, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component compatible with Java 11 and the consumer needed a component compatible with Java 8

rusheb commented on psi-sample/container-gradle:
```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env bash
```
Why include `container-gradle` rather than using it directly from the SDK?

snedamle replied:
done

rusheb commented on psi-sample/container-id:
```diff
@@ -0,0 +1 @@
+1d7a3c111e2b6f642861bce99940f9276970dfb9d498dcd2d9fb691c7cf38179
```
I don't think this file should be committed

snedamle replied:
removed

rusheb commented on psi-sample/enclave/build.gradle:
```diff
@@ -0,0 +1,42 @@
+        //
+        // Finally build the signed enclave:
+        //  ./gradlew build -PenclaveMode="Release"
+        //
```
Do we need these comments? I think they are copied directly from the hello world sample. In fact, do we need to use external signing at all or can we just use the private key?

snedamle replied:
removed all the signing configs.

rusheb commented on psi-sample/enclave/build.gradle:
```diff
@@ -0,0 +1,42 @@
+    debug {
+        signingType = privateKey
+        signingKey = file("../signing/sample_private_key.pem")
+    }
```
I think maybe we should choose to use dummy key or private key for both sim and debug mode. If we use dummy key, we should remove these blocks, including the commented code

snedamle replied:
removed all signing configs. I have kept debug, so that the users know how to use an external key for signing.

rusheb commented on psi-sample/enclave/build.gradle:
```diff
@@ -0,0 +1,42 @@
+}
+
+dependencies {
+    implementation "com.r3.conclave:conclave-enclave"
+    testImplementation "com.r3.conclave:conclave-testing"
+    testImplementation "org.junit.jupiter:junit-jupiter:5.6.0"
+    implementation project(":common");
```
for consistency with the client module I think it would be better to put the dependency on `common` first. I also think it's nice to separate the test dependencies
```suggestion
    implementation project(":common");

    implementation "com.r3.conclave:conclave-enclave"
    
    testImplementation "com.r3.conclave:conclave-testing"
    testImplementation "org.junit.jupiter:junit-jupiter:5.6.0"
```

snedamle replied:
done

rusheb commented on psi-sample/enclave/build.gradle:
```diff
@@ -0,0 +1,42 @@
+
+dependencies {
+    implementation "com.r3.conclave:conclave-enclave"
+    testImplementation "com.r3.conclave:conclave-testing"
```
Just FYI you will have to remove this when you migrate the sample to Conclave 1.1

snedamle replied:
yes

rusheb commented on psi-sample/gradle.properties:
```diff
@@ -0,0 +1,5 @@
+#
```
This file should not be committed

snedamle replied:
The sample requires this file to point to conclave sdk location.

rusheb commented on psi-sample/host/build.gradle:
```diff
@@ -0,0 +1,30 @@
+// Create a task that can be used for generating signing materials
+tasks.register("prepareForSigning") {
+    it.dependsOn(":enclave:generateEnclaveSigningMaterial${mode.capitalize()}")
+}
```
If we decide not to use external signing, we can also remove this task

snedamle replied:
removed.

rusheb commented on psi-sample/host/build.gradle:
```diff
@@ -0,0 +1,30 @@
+        outputs.upToDateWhen { false }
+        showStandardStreams = true
+    }
+}
```
Since this is a sample, would it worth adding comment explaining what this does?

snedamle replied:
removed this as this is not required here.

rusheb commented on psi-sample/signing/external_signing_public.pem:
```diff
@@ -0,0 +1,11 @@
+-----BEGIN PUBLIC KEY-----
```
If we decide not to use external signing we can delete this file

snedamle replied:
done

rusheb commented on psi-sample/signing/sample_private_key.pem:
```diff
@@ -0,0 +1,39 @@
+-----BEGIN RSA PRIVATE KEY-----
```
If we decide not to use external signing we can delete this file

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+## PSI - Private Set Intersection Sample
+
+This is a simple app using the Conclave API. It is licensed under the Apache 2 license, and therefore you may 
+copy/paste it to act as the basis of your own commercial or open source apps.
```
Do we want to add a link here to the docsite or to the conclave download page?

snedamle replied:
done

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+## PSI - Private Set Intersection Sample
```
I'm not sure what your convention is but this might be a nicer title rather than starting with an acronym
```suggestion
## Conclave Sample: Private Set Intersection
```

snedamle replied:
done

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+
+Using this sample will see how Conclave (based on Intel SGX) can be a new tool to solve the private set intersection problem (PSI)
+
+# What is Private Set Intersection Problem (PSI) ?
```
```suggestion
# What is Private Set Intersection (PSI)?
```

snedamle replied:
done

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+PSI refers to the problem of determining the common elements from the intersection of two sets without leaking or disclosing any 
+additional information of the remaining elements of the either sets.
+
+# Use Case : Measuring ad conversion rates
```
```suggestion
# Use case : measuring ad conversion rates
```

snedamle replied:
done

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+
+Using this sample will see how Conclave (based on Intel SGX) can be a new tool to solve the private set intersection problem (PSI)
+
+# What is Private Set Intersection Problem (PSI) ?
```
Maybe you should move the link to your blog post to this section as it is most relevant here? Also it might be nice to add an extra example and/or an image. However, might be better to keep it short and sweet as you have. Not sure.

snedamle replied:
Added an image explaining the use case.

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+Measuring ad conversion rates is done by comparing the list of people who have seen an ad with those who have completed a transaction. 
+These lists are held by the advertiser (say, Google or Facebook), and by merchants, respectively. 
+It is often possible to identify users on both ends, using identifiers such as credit card numbers, email addresses, etc. 
+A simple solution, which ignores privacy, is for one side to disclose its list of customers to the other side, which then computes the necessary statistics. 
```
I had to read this a few times to understand the use case. Not sure how we can make it more clear, but maybe worth some thought

rusheb replied:
Maybe a diagram with conclave in the middle. many arrows into conclave from one side showing ad views A B C, arrows into conclave from the other side showing transactions A C D. Finally arrows going back to each party showing which of their transactions formed a conversion (A and C for each)?

rusheb replied:
Having read the rest of the README the example has become quite clear! So maybe I should have been more patient. I think maybe it would be possible to reorganise the page using subheading levels

snedamle replied:
Added a diagram with conclave in the middle.

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+A simple solution, which ignores privacy, is for one side to disclose its list of customers to the other side, which then computes the necessary statistics. 
+
+# Parties involved in the sample
+Client1, Client2, Host.
```
Could we rename Client1 and Client2 to Merchant and ServiceProvider respectively?

snedamle replied:
done.

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+It is often possible to identify users on both ends, using identifiers such as credit card numbers, email addresses, etc. 
+A simple solution, which ignores privacy, is for one side to disclose its list of customers to the other side, which then computes the necessary statistics. 
+
+# Parties involved in the sample
```
I think this should be a subsection
```suggestion
## Parties involved in the sample
```

snedamle replied:
done.

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+
+# How to run
+
+Start the host on a non-Linux system, which will build the enclave and host:
```
We can also use normal gradle on a linux system, right?

snedamle replied:
Added steps to run on linux system using ./gradlew instead of ./container-gradle.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+ * This class is the Enclave class. To create an enclave in Conclave extend your class by Enclave interface.
+ * This class will contain the our business logic to calculate the ad conversion rate by using the list of credit card numbers given by the merchant
+ * and the service provider. Since the lists are shared with the enclave and not with anyone else, complete privacy of data is guaranteed.
+ */
```
Is the line length of these Javadocs ok? Not sure what the convention is but the lines overflow in the github renderer

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.stream.Collectors;
+
+/**
+ * This class is the Enclave class. To create an enclave in Conclave extend your class by Enclave interface.
```
I don't think you can extend an interface. `Enclave` is an abstract class.
```suggestion
 * This class is the Enclave class. To create an enclave in Conclave, use a class that extends Enclave.
```

snedamle replied:
done.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+
+/**
+ * This class is the Enclave class. To create an enclave in Conclave extend your class by Enclave interface.
+ * This class will contain the our business logic to calculate the ad conversion rate by using the list of credit card numbers given by the merchant
```
```suggestion
 * This class contains the business logic to calculate the ad conversion rate, using the list of credit card numbers given by the merchant
```

snedamle replied:
done.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.UUID;
+
+/**
+ * Client can be of two types- Merchant or Service Provider
```
```suggestion
 * Client can be of two types: Merchant or Service Provider
```

snedamle replied:
done.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+        if (args.length == 0) {
+            System.err.println("Please pass [MERCHANT/SERVICE-PROVIDER] followed by list of credit card numbers separated by spaces");
+            return;
+        }
```
Should we do this check right at the start to avoid creating the socket if the arguments are wrong? 
On the other hand, maybe it's more readable to put it here, so I'm not sure.

snedamle replied:
Moved this check to the top.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+
+        if("MERCHANT".equals(type)) {
+            userDetailsList = new ArrayList<>(args.length);
+            for (int i =2;i< args.length ; i++) {
```
are we missing out `args[1]`?

snedamle replied:
yes corrected this. args[1] will be the constraint.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+
+        if("MERCHANT".equals(type)) {
+            userDetailsList = new ArrayList<>(args.length);
+            for (int i =2;i< args.length ; i++) {
```
Whitespace
```suggestion
            for (int i =2; i < args.length; i++) {
```

snedamle replied:
done.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+
+        } else if("SERVICE-PROVIDER".equals(type)) {
+            adDetailsList = new ArrayList<>(args.length);
+            for (int i =2;i< args.length ; i++) {
```
Ditto

snedamle replied:
done.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+        fromHost.readFully(attestationBytes);
+
+        //STEP 4: Convert byte[] received from host to EnclaveInstanceInfo object
+        EnclaveInstanceInfo instanceInfo = EnclaveInstanceInfo.deserialize(attestationBytes);
```
we haven't actually checked the attestation against a constraint!

snedamle replied:
added EnclaveConstraint.parse("C:"+ constraint +" SEC:INSECURE").check(instanceInfo);

rusheb commented on psi-sample/common/src/main/java/com/r3/conclave/sample/common/InputData.java:
```diff
@@ -0,0 +1,37 @@
+        this.userDetailsList = userDetailsList;
+        this.adDetailsList = adDetailsList;
+    }
+
```
Is this used?

snedamle replied:
Yes it is used by InputDataSerializer when it de serializes the InputData object in the enclave class.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+        try {
+            EnclaveHost.checkPlatformSupportsEnclaves(true);
+        } catch (EnclaveLoadException e) {
+            e.printStackTrace();
```
I'm not sure if we want to print the stack trace here. If we run on macOS the method might throw, which could confuse users

snedamle replied:
Added a sysout instead.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            e.printStackTrace();
+        }
+
+        //loads enclave
```
The comment style is inconsistent - some comments start with lower case and some start with upper case

snedamle replied:
Made the comment style consistent throughout.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+    public static void main(String args[]) throws EnclaveLoadException, IOException {
+        final String enclaveClassName = "com.r3.conclave.sample.enclave.ReverseEnclave";
+
+        //Check if Platform supports enclave.
```
I don't know what this comment adds that is not described in the method name. We should remove it or add more details

snedamle replied:
removed this comment.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            e.printStackTrace();
+        }
+
+        //loads enclave
```
Same here - I don't know what this comment adds that is not described in the method name

snedamle replied:
Removed this comment.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            //accept is a blocking call and will wait for any client to connect
+            Socket clientSocket = acceptor.accept();
+
+            //routing hint identifies a particular client. This is used by enclave to send reply back to host
```
What do you mean "used by enclave to send reply back to host"?

snedamle replied:
Updated the comment.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            //routing hint identifies a particular client. This is used by enclave to send reply back to host
+            String routingHint = UUID.randomUUID().toString();
+
+            //clientMap used to maintain routing hint and client socket. This can be used by the host to send reply back to a particular client socket using routing hint
```
Maybe good to split this comment onto two lines. Same for some of the other long ones. I won't comment on all of them

snedamle replied:
Yes, have split all the longer comments on multiple lines.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+                            System.out.println("Host : Sending message to client");
+
+                            sendBytesToClient(((MailCommand.PostMail) mailCommand).getRoutingHint()
+                                    , ((MailCommand.PostMail) mailCommand).getEncryptedBytes());
```
Is the linter happy with this formatting?

snedamle replied:
My Sonar does not complain of this formatting.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+
+            //this uses DCAP attestation service and registers a callback which gets called when enclave wanst to send a reply back to client
+            enclaveHost.start(new AttestationParameters.DCAP() , mailCommands -> {
+                for(MailCommand mailCommand  : mailCommands) {
```
```suggestion
                for(MailCommand mailCommand : mailCommands) {
```

snedamle replied:
done.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+                            System.out.println("Host : Sending message to client");
+
+                            sendBytesToClient(((MailCommand.PostMail) mailCommand).getRoutingHint()
+                                    , ((MailCommand.PostMail) mailCommand).getEncryptedBytes());
```
Can we cast the `mailCommand` to PostMail outside this method call to improve readability?

snedamle replied:
done.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            //clientMap used to maintain routing hint and client socket. This can be used by the host to send reply back to a particular client socket using routing hint
+            clientMap.put(routingHint, clientSocket);
+
+            //this uses DCAP attestation service and registers a callback which gets called when enclave wanst to send a reply back to client
```
```suggestion
            //this uses DCAP attestation service and registers a callback which gets called when enclave wants to send a reply back to client
```

snedamle replied:
done.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            //clientMap used to maintain routing hint and client socket. This can be used by the host to send reply back to a particular client socket using routing hint
+            clientMap.put(routingHint, clientSocket);
+
+            //this uses DCAP attestation service and registers a callback which gets called when enclave wanst to send a reply back to client
```
I don't think we need to say that it uses DCAP attestation service here. Not that relevant to the sample.

It might be more helpful to describe exactly when the callback will be triggered in this application (i.e. after both the MERCHANT and SERVICE-PROVIDER have provided their data)

snedamle replied:
done.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+
+    /**
+     * This method gets called when client wants to communicate to enclave, and sends a message wrapped in a mail to host. Host in turn calls deliverMail method which in turn
+     * calls this method. In this method, we will deserialize the mail message, perform the computation and send the result back to the clients.
```
Comments in this file may also need separating into more lines

snedamle replied:
done.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+
+
+    /**
+     * This method gets called when client wants to communicate to enclave, and sends a message wrapped in a mail to host. Host in turn calls deliverMail method which in turn
```
inconsistent casing: host/client and Host/Client. Would be good to check throughout the repo and make sure it is consistent

```suggestion
     * This method gets called when client wants to communicate to enclave, and sends a message wrapped in a mail to host. Host in turn calls deliverMail method which in turn
```

snedamle replied:
Corrected this throughout teh project.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+            return;
+        }
+
+        String type = args[0];
```
Can we think of a better name than `type`? Maybe `clientType`?

snedamle replied:
renamed.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+        PostOffice postOffice = instanceInfo.createPostOffice(key, UUID.randomUUID().toString());
+
+        //STEP 6: Encrypt the message using enclave's public key
+        byte[] encryptedRequest = postOffice.encryptMail(serializeMessage(inputData).getBuffer(), type.getBytes());
```
So we are sharing the type of the client in the envelope which means that the host can view it? Is this ok for our application? Maybe we should add a caveat somewhere

snedamle replied:
I have added clientType to InputData. So no need to send client type via the envelope anymore.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+
+    public static void main(String args[]) throws InterruptedException, IOException, InvalidEnclaveException {
+
+        //STEP 1: Connect to Host Server
```
In the client file you use "STEP N: ..." style in the comments but not in the other files. What is the reason for the difference?

snedamle replied:
removed "STEP N" and made the comments consistent throughout the project.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+        toHost.writeInt(encryptedRequest.length);
+        toHost.write(encryptedRequest);
+
+        //STEP 8: Get the reply back from host via the socket
```
Maybe we should add a note/warning here that the host/enclave will not reply until both clients have sent a message.

Is there any risk of a timeout?

snedamle replied:
added a timeout of a minute.         acceptor.setSoTimeout(60000);

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+    private List<UserDetails> userDetailsList;
+    private List<AdDetails> adDetailsList;
+
+    private Map<String , String> routingHintMap = new HashMap();
```
maybe rename to `clientToRoutingHint` - for consistency and readability

snedamle replied:
done.

rusheb commented on psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,123 @@
+        PostOffice postOffice = instanceInfo.createPostOffice(key, UUID.randomUUID().toString());
+
+        //STEP 6: Encrypt the message using enclave's public key
+        byte[] encryptedRequest = postOffice.encryptMail(serializeMessage(inputData).getBuffer(), type.getBytes());
```
Also, maybe we should mention in the comment that we are adding the type in the envelope. We talk a lot about the envelope in the Enclave class but we don't mention it here, so it is not initially obvious where it comes from

snedamle replied:
moved clientType to InputData.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+ */
+public class PSIEnclave extends Enclave {
+
+    private static final String MERCHANT = "MERCHANT";
+    private static final String SERVICE_PROVIDER = "SERVICE-PROVIDER";
```
We use constants for these strings here but not in the client. Let's make this consistent. Maybe move the constants to `common`

snedamle replied:
Updated Client and made it consistent.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+        //deserialize the mail object using custom deserializers
+        InputData inputData = deserialize(mail);
+
+        //retrieve the envelope from mail. envelope is not encrypted
+        String envelope = new String(mail.getEnvelope());
```
I think we need to indicate that the contents of the envelope tell us what type of client it is

Maybe we could do this by adding a comment and renaming `envelope` to `clientType`

snedamle replied:
moved clientType to InputData.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+        //use clientToPublicKey which identifies the key of the client, which can be used by postoffice to encrypt data and send to client
+        clientToPublicKey.put(envelope, mail.getAuthenticatedSender());
+
+        //depending on the client type populate the two lists
```
```suggestion
        //depending on the client type populate one of the two lists
```

snedamle replied:
done.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        //once both the lists are populated, peform the ad conversion rate calculation
+        if(userDetailsList != null && adDetailsList != null && userDetailsList.size() > 0 && adDetailsList.size() > 0) {
```
Quite a lot of whitespace errors! Make sure to run a linter!

```suggestion
        if (userDetailsList != null && adDetailsList != null && userDetailsList.size() > 0 && adDetailsList.size() > 0) {
```

snedamle replied:
done.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+        if(userDetailsList != null && adDetailsList != null && userDetailsList.size() > 0 && adDetailsList.size() > 0) {
+            Double adConversionRate = getAdConversionRate(userDetailsList, adDetailsList);
+
+            //send the ad conversion rate to merchant. use merchants public key to encrypt the message such that only merchant will be able to decrypt it
```
```suggestion
            //send the ad conversion rate to merchant. use merchant's public key to encrypt the message such that only merchant will be able to decrypt it
```

snedamle replied:
done.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+            byte[] encryptedReply = postOffice(clientToPublicKey.get(MERCHANT)).encryptMail(adConversionRate.toString().getBytes());
+            postMail(encryptedReply, routingHintMap.get(MERCHANT));
+
+            //send the ad conversion rate to service provider. use service providers public key to encrypt the message such that only merchant will be able to decrypt it
```
```suggestion
            //send the ad conversion rate to service provider. use service provider's public key to encrypt the message such that only merchant will be able to decrypt it
```

snedamle replied:
done.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    /**
+     * Thsi calculates ad conversion rate.
```
```suggestion
     * This calculates ad conversion rate.
```

snedamle replied:
done.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            byte[] mailBytes = new byte[input.readInt()];
+            input.readFully(mailBytes);
+
+            //forward the request to enclave. the request is encrypted using enclave s public key and hence is not accessible to host
```
```suggestion
            //forward the request to enclave. The request is encrypted using enclave's public key and hence is not accessible to host
```

snedamle replied:
done.

rusheb commented on psi-sample/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,91 @@
+            Socket clientSocket = acceptor.accept();
+
+            //routing hint identifies a particular client. This is used by enclave to send reply back to host
+            String routingHint = UUID.randomUUID().toString();
```
If we do decide to keep the client type in the unencrypted envelope, we might as well use it for the routing hint rather than using a random UUID

snedamle replied:
It is a good idea to use client type as a routing hint if we would have kept client type in the envelope.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+ * This class will contain the our business logic to calculate the ad conversion rate by using the list of credit card numbers given by the merchant
+ * and the service provider. Since the lists are shared with the enclave and not with anyone else, complete privacy of data is guaranteed.
+ */
+public class PSIEnclave extends Enclave {
```
Do we need to add any tests?

snedamle replied:
Added MockTest to unit test enclave ad conversion logic.

rusheb commented on psi-sample/README.md:
```diff
@@ -0,0 +1,55 @@
+It is often possible to identify users on both ends, using identifiers such as credit card numbers, email addresses, etc. 
+A simple solution, which ignores privacy, is for one side to disclose its list of customers to the other side, which then computes the necessary statistics. 
+
+# Parties involved in the sample
```
Maybe it would be good to add a note on the best way to read the sample, e.g.
* focus on Enclave, Host, and Client classes
* what order to read them in?

snedamle replied:
Have updated the Readme to make it more clear.

rusheb commented on psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java:
```diff
@@ -0,0 +1,117 @@
+    @Override
+    protected void receiveMail(long id, EnclaveMail mail, String routingHint) {
+
+        //deserialize the mail object using custom deserializers
+        InputData inputData = deserialize(mail);
```
How difficult would it be to have serviceProviderData and merchantData as subclasses of inputData? That way the enclave could determine the type of the client based on the type of the InputData class and we wouldn't have to mess around with routing hints and envelopes

snedamle replied:
I have found a way around not using the default serialization/reflection as it is not supported in the current Graal Version. SO I am using Kryo which lets me define a de-serializer that can be used at the enclave side. Hence as of now I have added lists of serviceProviderData and merchantData to inputData, and have written serializers and deserialzers for inputData to get away with the serialization isses with current Graal VM. I will upgrade the sample to using serviceProviderData and merchantData as subclasses of inputData when we release 1.1.

rusheb commented on psi-sample/common/src/main/java/com/r3/conclave/sample/common/InputDataSerializer.java:
```diff
@@ -0,0 +1,61 @@
+package com.r3.conclave.sample.common;
```
I haven't used kryo before so I haven't looked at this file in detail.

snedamle replied:
I have written custom serializers and de serializers which is used by Kryo to serialize and deserialize the data from client to the enclave and vice versa.

rusheb commented on psi-sample/common/src/main/java/com/r3/conclave/sample/common/InputDataSerializer.java:
```diff
@@ -0,0 +1,61 @@
+package com.r3.conclave.sample.common;
```
How did you get kryo to work without having to specify the serialization/reflection config?

snedamle replied:
I have written custom serializers and de serializers which is used by Kryo to serialize and deserialize the data from client to the enclave and vice versa. Since Kryo uses this class for serialization/de serialization, reflection is also not required internally and hence enclave does not complian.

rusheb replied:
cool! nice workaround

rusheb commented:
There are quite a lot of grammar/formatting errors which I think are especially important to fix if this is going to be used by customers. I've left a number of comments wherever I noticed things but a more thorough review might be necessary.

On the whole I think this is a great sample and you have made the use case really clear and easy to understand. nice work!

rusheb reviewed (commented):
There are a lot of grammar/formatting errors which I think are especially important to fix if this is going to be used by customers. I've left a number of comments wherever I noticed things but a more thorough review might be necessary.

On the whole I think this is a great sample and you have made the use case really clear and easy to understand. nice work!

snedamle committed:
1. cleaned up the comments
2. adding .gitignore file
3. removing unnecessary files
4. adding client type to input data
5. Updating readme
6. Removing unnecessary tasks from build.gradle
7. Removing debug, release signing key from build

snedamle committed:
1. added a diagram illustrating psi

snedamle committed:
1. added a diagram illustrating psi

snedamle committed:
1. added a diagram illustrating psi

snedamle committed:
1. added a diagram illustrating psi

snedamle committed:
1. updating readme

snedamle committed:
Adding enclave unit test

snedamle committed:
1. updating image

snedamle committed:
1. updating readme

snedamle committed:
Delete .DS_Store

snedamle committed:
Delete .DS_Store

snedamle committed:
Delete .DS_Store

snedamle committed:
Delete .DS_Store

snedamle committed:
I will update README with blog link once it is published.

snedamle committed:
Merge branch 'psi' of https://github.com/R3Conclave/conclave-samples into psi
